Chat server

DavidGelmanSerg opened the pull request:
Приложение чат-сервер на TCP сокетах

DavidGelmanSerg committed:
ru.gelman: некоторые зависимости вынесены на общий уровень
chat-server: Создан новый модуль. Создана структура проекта. Созданы файлы конфигурации. Определены Entity. Определены операции над пользователями: создание, логин, получение пользователя. Определены необходимые исключения. Реализована операция создания сессии (после логина пользователя).
common: Создан новый модуль. Добавлен класс PropertyLoader.

DavidGelmanSerg committed:
entity:
ID сущностей больше не final.
Добавлена фабрика ChatEntityFactory.
Конструкторы каждой entity теперь package private и все entity создаются через ChatEntityFactory.
Изменена структура сущности Chat: информация о чате вынесена в ChatInfo.

service:
Добавлен функционал создания чата

mapper:
В ServiceMapper добавлены новые мапинги для entity
Все методы ServiceMapper теперь static
Добавлен класс JdbcRepositoryMapper для маппинга ResultSet на Entity

repository:
Методы save теперь возвращают тот тип, который получают.

DTO:
Добавлены необходимые DTO.

DavidGelmanSerg committed:
service:
Добавлен функционал отправки сообщения в чат

DavidGelmanSerg committed:
entity:
Изменена структура ChatSession

service:
Изменен алгоритм создания и сохранения сессии.
Изменен алгоритм проверки активности сессии

repository:
Изменена структура бд. Добавлена таблица Sessions для хранения клиентских сессий.
Хранение и получение сессий происходит не в Runtime, а через репозиторий.

DavidGelmanSerg committed:
service:
Добавлена поддержка активации сессии для обеспечения реконнекта клиентов.
dto:
Обновлен DTO SessionData

DavidGelmanSerg committed:
Изменена структура ChatMessage и MessageData - creatorId заменен на creator;
Конструкторы всех сущностей теперь имеют public;
Удален класс ChatEntityFactory. Его функционал перенесен в ServiceMapper;
Добавлены функции:
1. Получение активных сессий
2. Получение информации обо всех чатах пользователя
3. Получение сообщений чата
4. Получение всех пользователей чата
Добавлен метод маппинга ResultSet на ChatInfo в JdbcRepositoryMapper

DavidGelmanSerg committed:
Доработки логирования

DavidGelmanSerg committed:
Доработана связка контроллера и сервиса

DavidGelmanSerg committed:
Оптимизация запросов в БД

DavidGelmanSerg committed:
Отлажены запросы в Базу данных.
Изменены интерфейсы репозитория и сервиса для операции login

DavidGelmanSerg committed:
Добавлен метод получения активных сессий для чата по id
Добавлен core пакет для предоставления api сервера:
Настроен запуск приложения из .properties конфига
Изменен тип данных sessionId в запросе БД

Yuksovich commented on chat-server/build.gradle:
```diff
@@ -0,0 +1,28 @@
+    implementation 'com.fasterxml.jackson.core:jackson-databind:2.15.0'
+    implementation 'com.fasterxml.jackson.core:jackson-annotations:2.12.4'
+    implementation 'com.fasterxml.jackson.core:jackson-core:2.15.0'
+    implementation group: 'com.h2database', name: 'h2', version: '2.2.220'
```
старайся использовать единый формат. Потому как у тебя тут по сути 4 формата записи зависимости:
через одинарные кавычки, через двойные кавычки, через скобочки, без скобочек, слитно группа:имя:версия, и все поотдельности через group: ... name:... version:...
ну и лучше порядок делать наоборот, по степени важности так сказать, сначала основные зависимости, а тестовые в конце

DavidGelmanSerg replied:
Исправил

Yuksovich commented on chat-server/src/main/java/ru/gelman/ChatServerApplication.java:
```diff
@@ -0,0 +1,71 @@
+                    Runnable mainHandler = new ChatServerRequestHandler(handlers, sessionManager, clientSocket);
+                    handleThreads.submit(mainHandler);
+                }
+            } catch (IOException e) {
```
обрати внимание, что любая ошибка, приведет к выходу из цикла и, соответственно, остановке сервера

DavidGelmanSerg replied:
Исправил. При каждом IOException ошибка логируется и сокет открывается заново.

Yuksovich commented on chat-server/src/main/resources/query.properties:
```diff
@@ -0,0 +1,56 @@
+#CREATE
+create_users_table=CREATE TABLE IF NOT EXISTS Users(\
```
query в пропертях нет никакого смысла хранить, обычно скрипты для создания таблиц хранятся в файлах миграции и используются библиотеками наподобии liquibase или flyway.
В крайнем случае можно просто хранить sql файл с DDL скриптами.
DDL - это data definition language - это скрипт в которм создаются таблицы и определяется структура БД.
Второй тип - это DML - это data modification language - это твои запросы на вставку и модификацию данных (insert и update).

В общем им нет смысла храниться в проперти файлах потому, что они так просто не меняются, а в пропертях хранят какие-то настройки, которые можно быстро и легко поменять, но скрипт нельзя просто поменять, это не будет иметь особого смысла. Плюс хранить его в в sql файле удобно, потому, что ты можешь просто выполнить его на произвольной БД, если потребуется.

DavidGelmanSerg replied:
Исправил. Перевел все на liquibase

Yuksovich commented on chat-server/src/main/resources/query.properties:
```diff
@@ -0,0 +1,56 @@
+  FOREIGN KEY (userId) REFERENCES Users(id) ON DELETE CASCADE\
+  );
+#INSERT
+insert_user=INSERT INTO Users (name, password, deleted) VALUES (?, ?, ?);
```
запросы dml это по сути часть кода их не надо хранить отдельно, это неудобно и не несет каких-то профитов. Лучше сразу в коде видеть, что за запрос и какие у него параметры, плюс идея подсветит, если есть какие-то ошибки в нем.

DavidGelmanSerg replied:
Исправил

Yuksovich commented on chat-server/src/main/java/ru/gelman/repository/h2/ChatH2Repository.java:
```diff
@@ -0,0 +1,472 @@
+        password = config.getProperty("password", "");
+
+        try (Connection connection = getConnection(); Statement statement = connection.createStatement()) {
+            String createUsersTableQuery = getQuery("create_users_table");
```
а что будем делать, если я тебе скажу использовать нормальную БД, не in-memory? которая будет сохранять на диск данные? при рестарте приложения оно выкинет ошибку, что все эти таблицы уже есть

DavidGelmanSerg replied:
Видимо, это последствия ненаглядности кода. В моих SQL запросах встроена инструкция IF NOT EXISTS. Разве она не поддерживается большинством популярных СУБД как раз чтобы избежать дублирования таблиц? Можно было в целом использовать обработку через конфиг (когда в строке пишешь IFEXISTS=true) и exception, но я не знаю, есть ли такие фишки у того postgreSQL или Oracle там. Кажется, что это узконаправленная штука. Поэтому я решил использовать SQL инструкции.

DavidGelmanSerg replied:
Исправил. Перевел все на liquibase

Yuksovich commented on chat-server/src/main/java/ru/gelman/repository/ChatRepository.java:
```diff
@@ -0,0 +1,44 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public interface ChatRepository {
```
тут нарушены Interface segregation principle - разделение интерфейсов. Один интерфейс отвечает за все. И за пользователя и за чат и за сообщение и за сессию. Это стоит разнести в отдельные репозитории. UserRepository, SessionRepository ...

DavidGelmanSerg replied:
Исправил. Разделил все на три репозитория: Для юзеров, для сессий и для чатов. Хотел вынести репозиторий сообщений отдельно, но кажется в этом нет необходимости.

DavidGelmanSerg replied:
Разделил сервисы и контроллеры также

Yuksovich commented on chat-server/src/main/java/ru/gelman/repository/h2/ChatH2Repository.java:
```diff
@@ -0,0 +1,472 @@
+    @Override
+    public boolean login(ChatUser user) {
+        try (Connection connection = getConnection()) {
+            PreparedStatement loginQuery = connection.prepareStatement(getQuery("select_user_login"));
```
одна из причин почему не стоит хранить где-то далеко запросы в БД - потому, что я вот не понимаю так сходу, что же тут происходит на самом деле. Мне надо куда-то перепрыгнуть, чтобы найти, что же за запрос select_user_login.

DavidGelmanSerg replied:
Исправил. Перенес все запросы в код

Yuksovich commented on chat-server/src/main/java/ru/gelman/repository/h2/ChatH2Repository.java:
```diff
@@ -0,0 +1,472 @@
+
+    @Override
+    public ChatUser getUser(String name) {
+        try (Connection connection = getConnection()) {
```
заметь, вот этот 
try (Connection connection = getConnection()) {
connetcion prepareStatement
}catch (SQLException e)..
повторяется постоянно, может стоит как-то красиво это обыграть, вынести в отдельный метод? Маленькая подсказка - использовать лямбда выражения

DavidGelmanSerg replied:
Исправил

Yuksovich commented on chat-server/src/main/java/ru/gelman/mapper/ServiceMapper.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class ServiceMapper {
```
ознакомься с либой MapStruct

DavidGelmanSerg replied:
Исправил

Yuksovich commented on chat-server/src/main/java/ru/gelman/entity/Chat.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Override
+    public int compareTo(Chat o) {
```
эта логика в целом не очень понятна, сравнения чатов по последнему сообщению - это лучше вынести в какой-то отдельный сервис или компаратор, который названием будет говорить, что он сравнивает. А так это выглядит очень странно

DavidGelmanSerg replied:
Исправил. Вынес в отдельный компаратор.

Yuksovich commented on chat-server/src/main/java/ru/gelman/entity/ChatMessage.java:
```diff
@@ -0,0 +1,34 @@
+@ToString(onlyExplicitlyIncluded = true)
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
+@AllArgsConstructor
+public class ChatMessage implements Comparable<ChatMessage> {
```
обычно не делают такую страшную портянку из аннотаций. Очень сложно разобраться, что же тут происходит и почему так много каких-то исключений. Возможно стоит просто поставить @Data

DavidGelmanSerg replied:
Исправил. Все ентити теперь @Data

Yuksovich commented on chat-server/src/main/java/ru/gelman/core/response/ChatResponse.java:
```diff
@@ -0,0 +1,107 @@
+import java.util.Map;
+
+public class ChatResponse {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
```
маппер лучше отдельно сконфигурировать и подсовывать его всем, кому нужен. Он обычно является синглтоном, он потокобезопасен и конфигурируется отдельно. Не место ему тут конфигурироваться

DavidGelmanSerg replied:
Сделал

Yuksovich commented on chat-server/src/main/java/ru/gelman/core/client/TcpClient.java:
```diff
@@ -0,0 +1,86 @@
+                 var out = new PrintWriter(new OutputStreamWriter(client.getOutputStream(), StandardCharsets.UTF_8), true);
+                 var in = new Scanner(client.getInputStream(), StandardCharsets.UTF_8)
+            ) {
+                String json = chatResponse.toString();
```
toString не надо использовать в логике, сделай отдельный метод. toString это просто для какой-нибудь отладки или логирования и должен работать быстро. И уж точно он не должен вдруг выкидывать исключения.

DavidGelmanSerg replied:
Исправил. Вынес в отдельный метод

Yuksovich commented on chat-server/src/main/java/ru/gelman/ChatServerApplication.java:
```diff
@@ -0,0 +1,71 @@
+            ChatController controller = new ChatController(service);
+
+            log.info("initializing handlers");
+            Map<ChatAction, ChatRequestHandler> handlers = new HashMap<>();
```
а сделай через рефлексию. Не так, что надо вручную добавлять кучу хендлеров, а чтобы каждый новый хендлер добавлялся автоматически, если просто написать новую реализацию ChatRequestHandler. Open closed principle так сказать

DavidGelmanSerg replied:
Выполнил

DavidGelmanSerg committed:
Создан пакет network в модуле common. В него помещены общие абстракции для общения в сети (Интерфейсы Client и NetMessage);
Создан модуль chat-client - простейший консольный чат-клиент;
Хендлеры запросов в модуле chat-server переведены на использование пакета network (см. п.1);
Конструкторы хендлеров запросов в модуле chat-server стали без аргументов;
Контроллеры сервисов перемещены в контекст запроса;
В модуль chat-server добавлены зависимости liquibase, создана привязка к задаче run.
Все entity в модуле chat-server помечены аннотацией @Data; добавлена entity ChatInfo
Контроллеры, сервисы и репозитории разделены согласно зонам ответственности
В модуле chat-server добавлена рефлексивная загрузка обработчиков на запуске приложения
Переработана механика хендлеров и их фильтрации.
Добавлена сущность Sender для инкапсуляции логики рассылки исходящих событий
Реализован перезапуск сокета сервера при ошибках
Реализован цикл общения с подключаемыми клиентами, пока сокет открыт
В entity классы добавлены статические методы для читаемого создания сущностей
Удален класс маппера сущностей. Маппер создается через MapStruct
в модуле common создан пакет обмена сообщениями, включающий MessageStore, MessageConsumer, MessageProducer
Удалены неиспользуемые классы

DavidGelmanSerg committed:
1. Синхронизировал методы в TcpClient
2. Поменял объявление мапы в ClientManager.java
3. Дописал help.txt

DavidGelmanSerg committed:
Добавлены недостающие файлы

DavidGelmanSerg committed:
Синхронизирован метод getResponse в TcpClient.java, так как является составным действием.

DavidGelmanSerg committed:
Методы класса TcpClient.java сделаны несинхронизированными, так как в этом нет смысла, потоки сокета автоматически блокируют поток исполнения при проведении операции ввода-вывода.
